### DISCUSSION TITLE
How do I open a saved PDF file on my Android device ?

### DISCUSSION THREAD

**Discussion by leonelgregioauler** [apache/cordova#510] [Q&A] [answered]
Hello everyone !

I'm generating a PDF file using jsPDF library. After that, I've been using cordova-plugin-file to write it into an Android folder. The file are been writing sucessfull ! Even into "dataDirectory" (internal) or "externalRootDirectory" (external)

The problem is : 

I've tried to open this file, by several ways, using:

- The library command "doc.save('fileName.pdf')".
- window.open
- cordova.InAppBrowser.open
           This 2 commands above, using the physical file Path:
                window.open(filePath, '_blank');
            Or ....
            blob content as described below:
              const urlBlob = URL.createObjectURL(blob);
              window.open(urlBlob, '_blank');
 
- Tag "embed"
- And many other ways that I don't remember..... 

It seems to be something like a permission
None of this ways are effective in solve my need. Anyone has any solution ?

Thanks in advance !!!

**Comment by breautek** [comment, marked as answer]
PDFs are complex binaries, so a pdf viewer software is necessary.

The proper "android" way is to use an Intent to open the PDF document, which the user can choose a supported application of their choice (e.g. Adobe Acrobat Reader).

Cordova doesn't have an out-of-the-box solution to open intents in a general sense. A cordova plugin will need to expose a way to request for the PDF to be opened using an app capable of reading PDFs.

Alternatively, Mozilla as a PDF reader implementation in pure JS using an HTML canvas (https://mozilla.github.io/pdf.js/). This will likely be not very performant if I were to guess, but it could be good enough for your needs and probably worth looking into. The [getDocument](https://mozilla.github.io/pdf.js/api/) API supports data types like an ArrayBuffer so you'd just use the cordova file plugin to read the data as an `ArrayBuffer` then you should be able to pass the ArrayBuffer directly into mozzila's PDF library.

```javascript
function readBinaryFile(fileEntry) {

    fileEntry.file(function (file) {
        var reader = new FileReader();

        reader.onloadend = function() {
            let data = this.result; // This is the ArrayBuffer
        };

        reader.readAsArrayBuffer(file);

    }, onErrorReadFile);
}
```

**Reply by leonelgregioauler** [reply]
Thank you for the prompt response !!

Wow, it seems to be great ! I'll try !!

Thank you !

**Reply by leonelgregioauler** [reply]
Good morning ! 

Today, I did the implementation. 
Do you know if there is a way to "share" this PDF ? Or something like this. 
It can be "share" using the PDF.js or  "share" using the physical URL of the generated file, sending by whats app, e-mail.... 

![image](https://github.com/user-attachments/assets/2c63309f-a266-4406-b103-09fd76a4b533)

**Reply by breautek** [reply]
Kind of outside my expertise so I can't really explain how very well.

But the `dataDirectory` is app-private. It doesn't mean you can't share the files but other apps can't just read it. There's probably an android intent (which you'll need to find a plugin that exposes an API to use) for sharing files. This is typical for apps that want to have a share button to bring up some UI to let the user select which app they want to share the file with (e.g. email client, the clipboard, facebook, so on....)

The external storage partition apps also has slightly more flexibility, where you could store the file inside `Documents` folder but the file plugin cannot write to this directory. Modern Android (API 29+) limits the ability for the File APIs to read/write documents onto the external storage. Instead you must use a Media Store API, which there are also third-party plugins for. Once written to external storage, other apps can use file picking APIs where the user should be able to see and select a file to be opened in another app). Depending on your use case, this might be fine.

If you can't find a plugin that exposes file sharing and want to tackle it yourself by creating your own plugin, then https://developer.android.com/training/secure-file-sharing will probably be a start to understand the android API and https://cordova.apache.org/docs/en/12.x/guide/hybrid/plugins/index.html for creating a cordova plugin.

> or "share" using the physical URL of the generated file

To be clear on this, `file://` uris are definitely not shareable, especially if you're using internal storage (like the `dataDirectory`). External storage paths are shareable but the apps will need to have additional permissions, so it's not ideal and likely won't work the way you'd expect.

The native way to share files is to produce a `content://` uri for it, which can be shared with other apps. This is generally handled by the intent system.

**Reply by leonelgregioauler** [reply]
I've used the Media Store API plugin before. And it works. My only concern is that it's an "unofficial" Cordova plugin. And I have doubts about how it will work in future Android versions.

Otherwise, it works well for me in another application and I think it's the solution for what I need now.
